File: backend/src/main/java/br/com/genesis/ranking/service/ScoringService.java
package br.com.genesis.ranking.service;

import java.util.List;

import org.springframework.stereotype.Service;

import br.com.genesis.ranking.model.AthleteHistory;
import br.com.genesis.ranking.model.enums.HistoryType;

@Service
public class ScoringService {
  public static final int WIN_POINTS = 0;
  public static final int PODIUM_1ST = 3;
  public static final int PODIUM_2ND = 2;
  public static final int PODIUM_3RD = 1;

  public int calculateTotalPoints(List<AthleteHistory> history) {
    if (history == null) return 0;
    int total = 0;
    for (AthleteHistory record : history) {
      if (record == null || record.getType() == null) continue;
      if (record.getType() == HistoryType.WIN) total += WIN_POINTS;
      if (record.getType() == HistoryType.SEED && record.getPoints() != null) total += record.getPoints();
      if (record.getType() == HistoryType.PODIUM) {
        if (record.getPosition() != null) {
          if (record.getPosition() == 1) total += PODIUM_1ST;
          if (record.getPosition() == 2) total += PODIUM_2ND;
          if (record.getPosition() == 3) total += PODIUM_3RD;
        }
      }
    }
    return total;
  }

  public ScoreBreakdown buildBreakdown(List<AthleteHistory> history) {
    ScoreBreakdown breakdown = new ScoreBreakdown();
    if (history == null) return breakdown;

    for (AthleteHistory record : history) {
      if (record == null || record.getType() == null) continue;
      if (record.getType() == HistoryType.WIN) breakdown.wins += 1;
      if (record.getType() == HistoryType.SEED && record.getPoints() != null) {
        breakdown.seedPoints += record.getPoints();
      }
      if (record.getType() == HistoryType.PODIUM) {
        if (record.getPosition() != null) {
          if (record.getPosition() == 1) breakdown.podium1 += 1;
          if (record.getPosition() == 2) breakdown.podium2 += 1;
          if (record.getPosition() == 3) breakdown.podium3 += 1;
        }
      }
    }

    breakdown.total = calculateTotalPoints(history);
    return breakdown;
  }

  public static class ScoreBreakdown {
    public int total = 0;
    public int wins = 0;
    public int podium1 = 0;
    public int podium2 = 0;
    public int podium3 = 0;
    public int seedPoints = 0;
  }
}

File: src/services/scoringService.js
// Points configuration
export const SCORE_RULES = {
  WIN: 0,
  PODIUM_1ST: 3,
  PODIUM_2ND: 2,
  PODIUM_3RD: 1,
};

/**
 * Calculates total points for an athlete based on their history
 * @param {Array} history - List of results for the athlete
 * @returns {number} Total points
 */
export const calculateTotalPoints = (history) => {
  return history.reduce((total, record) => {
    let points = 0;
    if (record.type === 'win') points += SCORE_RULES.WIN;
    if (record.type === 'seed' && Number.isFinite(record.points)) points += record.points;
    if (record.type === 'podium') {
      if (record.position === 1) points += SCORE_RULES.PODIUM_1ST;
      if (record.position === 2) points += SCORE_RULES.PODIUM_2ND;
      if (record.position === 3) points += SCORE_RULES.PODIUM_3RD;
    }
    return total + points;
  }, 0);
};

/**
 * Builds a transparent breakdown of how the total score was composed
 * @param {Array} history
 * @returns {{ total: number, wins: number, podium1: number, podium2: number, podium3: number, seedPoints: number }}
 */
export const buildScoreBreakdown = (history) => {
  const breakdown = {
    total: 0,
    wins: 0,
    podium1: 0,
    podium2: 0,
    podium3: 0,
    seedPoints: 0,
  };

  history.forEach((record) => {
    if (record.type === 'win') breakdown.wins += 1;
    if (record.type === 'seed' && Number.isFinite(record.points)) breakdown.seedPoints += record.points;
    if (record.type === 'podium') {
      if (record.position === 1) breakdown.podium1 += 1;
      if (record.position === 2) breakdown.podium2 += 1;
      if (record.position === 3) breakdown.podium3 += 1;
    }
  });

  breakdown.total = calculateTotalPoints(history);
  return breakdown;
};

const normalizeName = (value) => (
  (value || '').toString().trim().toLowerCase()
);

const buildRankMeta = (athlete) => {
  const history = Array.isArray(athlete?.historico) ? athlete.historico : [];
  const breakdown = buildScoreBreakdown(history);

  return {
    pontos: Number(athlete?.pontos) || 0,
    podium1: breakdown.podium1,
    wins: breakdown.wins,
    name: normalizeName(athlete?.nome),
  };
};

const compareRankMeta = (a, b) => {
  if (a.pontos !== b.pontos) return b.pontos - a.pontos;
  if (a.podium1 !== b.podium1) return b.podium1 - a.podium1;
  if (a.wins !== b.wins) return b.wins - a.wins;
  return a.name.localeCompare(b.name);
};

/**
 * Ranks athletes by total points
 * @param {Array} athletes 
 * @returns {Array} Sorted athletes
 */
export const rankAthletes = (athletes) => {
  const metaCache = new Map();
  const getMeta = (athlete) => {
    if (!metaCache.has(athlete)) {
      metaCache.set(athlete, buildRankMeta(athlete));
    }
    return metaCache.get(athlete);
  };

  return [...athletes].sort((a, b) => compareRankMeta(getMeta(a), getMeta(b)));
};
